source/Nevermore/Util/DataModificationQueryBuilder.cs

```diff
@@ -330,87 +331,153 @@ CommandParameterValues GetDocumentParameters(Func<DocumentMap, string> allocateI
+                            parameters.Add(relatedVariableName, relatedDocumentId);
+                            if (x > 0)
+                                insertStatementBuilder.Append(",");
+                            insertStatementBuilder.AppendLine(
+                                $"(@{parentIdVariable}, '{mapping.TableName}', @{relatedVariableName}, '{relatedTableName}')");
+                            remaining--;
+                        }
+
+                        if (parameters.Count >= SqlServerParameterLimit && remaining > 0)
+                        {
+                            commands.Add(new PreparedCommand(insertStatementBuilder.ToString(), parameters, RetriableOperation.Insert, mapping, options.CommandTimeout));
+                            insertStatementBuilder = new StringBuilder();
+                            parameters = new CommandParameterValues();
+                            parameters.AddRange(documentParametersCreator());
+                        }
+                    }
                 }
             }
+
+            commands.Add(new PreparedCommand(insertStatementBuilder.ToString(), parameters, RetriableOperation.Insert, mapping, options.CommandTimeout));
+            return commands.ToArray();
         }
 
-        string AppendRelatedDocumentStatementsForUpdate(
-            string statement,
-            CommandParameterValues parameters,
+        PreparedCommand[] CreateUpdateCommandsWithRelatedDocuments(
+            string updateStatement,
             DocumentMap mapping,
-            object document)
+            object document,
+            UpdateOptions updateOptions)
         {
+            Func<CommandParameterValues> parametersCreator = () => GetDocumentParameters(
+                m => throw new Exception("Cannot update a document if it does not have an ID"),
+                null,
+                null,
+                document,
+                mapping,
+                DataModification.Update
+            );
+            var commands = new List<PreparedCommand>();
             var relatedDocumentData = GetRelatedDocumentTableData(mapping, new[] {document});
+            var parameters = parametersCreator();
+
             if (relatedDocumentData.Count == 0)
-                return statement;
+            {
+                commands.Add(new PreparedCommand(updateStatement, parameters, RetriableOperation.Update, mapping, updateOptions.CommandTimeout));
+                return commands.ToArray();
+            }
 
             var sb = new StringBuilder();
-            sb.AppendLine(statement);
+            sb.AppendLine(updateStatement);
             sb.AppendLine();
 
+            // sql commands are executed inside sp_executesql which means table variables and local temp tables won't live between commands
+            var globalTempTableName = $"##{configuration.RelatedDocumentsGlobalTempTableNameGenerator()}";
```

Comment:
This is the best solution I could come up with to allow this update approach to work over multiple commands.

Global temp tables are effectively reference counted and only exist while there are sessions open with references to the global temp table, so cleanup is automatically taken care of.

Reply:
An option to be consistent with the previous version is to generate a different SQL statement here if we only need to create a single command.
If we only need a single command we can use a table variable, otherwise we need the global temp table.